src/WebJobs.Script/Host/ScriptHostManager.cs

```diff
@@ -170,17 +170,17 @@ public void RunAndBlock(CancellationToken cancellationToken = default(Cancellati
 
                     newInstance = _scriptHostFactory.Create(_environment, EventManager, _settingsManager, _config);
                     _currentInstance = newInstance;
-                    lock (_liveInstances)
-                    {
-                        _liveInstances.Add(newInstance);
-                        _hostStartCount++;
-                    }
 
                     newInstance.HostInitializing += OnHostInitializing;
                     newInstance.HostInitialized += OnHostInitialized;
                     newInstance.HostStarted += OnHostStarted;
                     newInstance.Initialize();
 
+                    lock (_liveInstances)
+                    {
+                        _liveInstances.Add(newInstance);
+                    }
+
                     newInstance.StartAsync(cancellationToken).GetAwaiter().GetResult();
```

Comment:
It actually seems like the call to `StartAsync()` should happen before adding to `_liveInstances` as well. There's still a race here where the host can be disposed before we call `StartAsync()`, and it looks like we'll still orphan the instance in the exception handler if it throws. 

Thoughts?

Reply:
What was the issue with the initialization done in the factory? (perhaps @mathewc would know) It does feel like we should be initializing before adding to this collection, or, we need to better handle the disposal path when done during initialization.

Reply:
I believe that he wanted to hook into the host events before calling Initialize() and he didn't have a place to do that, so he broke the factory up a bit. Here's his comment: https://github.com/Azure/azure-functions-host/pull/2156#pullrequestreview-79720915